### PR TITLE
ROX-20879: Central-Sensor use hello for communicating deduper state

### PR DESCRIPTION
**Pull request by fredrb**
## Description

This PR fixes a few things as a prefactor for falling back the reconciliation if Central fails ([ROX-20878](https://issues.redhat.com/browse/ROX-20878))

1. Always send compatibilities on the connection. Compatibility should be used as an agreement that both sides are _able_ to exchange deduper messages. Not that either side will do so. 
2. Introduce a new property in both `SensorHello` and `CentralHello` to communicate intent of sending the deduper state
3. Move the feature flag to Central only. If the FF is off, Central will communicate to sensor (via the hello message) that deduper state shouldn't be sent.
4. (fix) Sensor should not send `unchangedIDs` in the ResourcesSynced message if client reconciliation is disabled. 

## Checklist
- [x] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- [ ] ~~Evaluated and added CHANGELOG entry if required~~
- [ ] ~~Determined and documented upgrade steps~~
- [ ] ~~Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~~

If any of these don't apply, please comment below.

### Here I tell how I validated my change

- [x] Unit tests
- [x] Integration tests
- [ ] Compatibility tests

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

### CODE REVIEW

File: sensor/common/sensor/central_communication_impl.go [hunk 1]
```diff
@@ -246,6 +244,16 @@ func (s *centralCommunicationImpl) initialSync(stream central.SensorService_Comm
 	centralCaps := centralHello.GetCapabilities()
 	centralcaps.Set(sliceutils.FromStringSlice[centralsensor.CentralCapability](centralCaps...))
 
+	// Sensor should only communicate deduper states if central is able to do so and it has requested it.
+	s.clientReconcile = s.clientReconcile &&
+		centralcaps.Has(centralsensor.SensorReconciliationOnReconnect) &&
+		centralHello.SendDeduperState
```
Comment (lvalerom) [inline sensor/common/sensor/central_communication_impl.go:250]:
```suggestion
		centralHello.GetSendDeduperState()
```

File: sensor/common/sensor/central_communication_impl.go [hunk 2]
```diff
@@ -246,6 +244,16 @@ func (s *centralCommunicationImpl) initialSync(stream central.SensorService_Comm
...
+
+	log.Infof("Sensor client reconciliation state=%s (centralCapability=%s, centralHello.SendDeduperState=%s)",
+		strconv.FormatBool(s.clientReconcile),
+		strconv.FormatBool(centralcaps.Has(centralsensor.SensorReconciliationOnReconnect)),
+		strconv.FormatBool(centralHello.SendDeduperState))
```
Comment (lvalerom) [inline sensor/common/sensor/central_communication_impl.go:255]:
```suggestion
		strconv.FormatBool(centralHello.GetSendDeduperState()))
```

File: sensor/common/sensor/sensor.go [hunk 3]
```diff
@@ -383,10 +383,7 @@ func (s *Sensor) communicationWithCentralWithRetries(centralReachable *concurren
 	exponential.InitialInterval = env.ConnectionRetryInitialInterval.DurationSetting()
 	exponential.MaxInterval = env.ConnectionRetryMaxInterval.DurationSetting()
 
-	if features.SensorReconciliationOnReconnect.Enabled() {
-		s.reconcile.Store(true)
-	}
-
+	s.reconcile.Store(true)
```
Comment (lvalerom) [inline sensor/common/sensor/sensor.go:386]: Same comment about the FF here.

File: sensor/common/sensor/central_communication_impl.go [hunk 4]
```diff
@@ -139,17 +140,14 @@ func (s *centralCommunicationImpl) sendEvents(client central.SensorServiceClient
...
 	for _, component := range s.components {
 		capsSet.AddAll(component.Capabilities()...)
 	}
-	if s.clientReconcile {
-		log.Info("Sensor is capable of doing client reconciliation")
-		capsSet.Add(centralsensor.SensorReconciliationOnReconnect)
-	} else {
-		log.Info("Sensor has client reconciliation disabled")
-	}
+	capsSet.Add(centralsensor.SensorReconciliationOnReconnect)
```
Comment (lvalerom) [inline sensor/common/sensor/central_communication_impl.go:150]: Not sure if it would be better to guard this with the FF

**Review by lvalerom — APPROVED** [review]
The change looks good. 

I'm not sure how I feel about setting the capability and the `reconcile` variable in Sensor unconditionally and not guarding them behind the FF. I believe that since the FF is checked in Central, and Central capability is also guarded by it, this won't be a problem, but I find it confusing to read.

**Comment by fredrb** [comment]
> I'm not sure how I feel about setting the capability and the reconcile variable in Sensor unconditionally and not guarding them behind the FF. I believe that since the FF is checked in Central, and Central capability is also guarded by it, this won't be a problem, but I find it confusing to read.

My problem is if we control in both central and sensor, then it becomes a double-switch for a single feature. Both of them have to be on for the feature to work. This is annoying for testing and it's annoying for any customers that would need to enable/disable this. 

Imagine this is disabled an user wants to enable it. They need to set an environment variable for _each_ secured cluster. If there are 60 secured clusters then this means updating 60 yaml files. With a centralized switch in central, enabling the feature means just enabling the env in central.

File: sensor/common/sensor/sensor.go (hunk 3 above)
Comment (fredrb) [reply sensor/common/sensor/sensor.go:386]: See comment below

File: sensor/common/sensor/central_communication_impl.go (hunk 4 above)
Comment (fredrb) [reply sensor/common/sensor/central_communication_impl.go:150]: See comment below

**Pull request merged**